tsdb/querier.go

```diff
@@ -148,6 +158,25 @@ func selectSeriesSet(ctx context.Context, sortSeries bool, hints *storage.Select
 	return newBlockSeriesSet(index, chunks, tombstones, p, mint, maxt, disableTrimming)
 }
 
+// truncatePostingsToLimit truncates the postings returned by p to the limit.
+func truncatePostingsToLimit(limit int, p index.Postings) index.Postings {
+	if limit <= 0 {
+		return p
+	}
+
+	out := make([]storage.SeriesRef, 0, 128)
+
+	for p.Next() {
+		id := p.At()
+		out = append(out, id)
+		if len(out) >= limit {
+			break
+		}
+	}
+
+	return index.NewListPostings(out)
```

Comment:
I'm concerned about the need to expand the postings iterator `p`, to implement truncation. Wouldn't it be more natural to wrap `p` with a limiting iterator? I don't remember if there already exists such an iterator type. WDYT?

Reply:
Maybe something like this (not tested fully)?

```
// LimitPostings limits the postings to limit.
type LimitPostings struct {
	p     Postings
	limit int
	i     int
}

// NewLimitPostings returns a Postings that can will iterate up to the limit. 0 means limit is disabled.
func NewLimitPostings(p Postings, limit int) Postings {
	if limit <= 0 {
		return p
	}
	return &LimitPostings{p: p, limit: limit}
}

func (it *LimitPostings) At() storage.SeriesRef {
	return it.p.At()
}

func (it *LimitPostings) Next() bool {
	if it.i > it.limit {
		return false
	}
	it.i++
	return it.p.Next()
}

func (it *LimitPostings) Seek(x storage.SeriesRef) bool {
	// If the current value satisfies, then return.
	if it.p.At() >= x {
		return true
	}

	for it.Next() {
		if it.At() >= x {
			return true
		}
	}

	return false
}
```

One trade off was that in `Seek()` each posting must be iterated one by one so that `i` can be kept track of. 
The `ListPostings` implementation of `Seek()` does binary search.
Trade off here is that Seek() becomes O(limit).

Reply:
I think the suggestion makes sense. Please take another look.